### PR TITLE
style: Perform variable substitution less often.

### PR DESCRIPTION
**Pull request by emilio**
We're wasting a bunch of work unneedlessly.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/servo/19370)

### CODE REVIEW

File: components/style/properties/properties.mako.rs [hunk 1]
```diff
@@ -3275,14 +3275,6 @@ where
         rule_cache_conditions: RefCell::new(rule_cache_conditions),
     };
 
-    let ignore_colors = !device.use_document_colors();
```
Comment (upsuper) [inline components/style/properties/properties.mako.rs]: Why are you removing this? It doesn't seem to me invoking `use_document_colors` is as cheap as checking a variable. Maybe optimizer can extract that as a loop invariant, but this seems to be unrelated to this commit anyway.

File: components/style/properties/properties.mako.rs [hunk 2]
```diff
@@ -3336,10 +3314,42 @@ where
...
+                        context.builder.custom_properties.as_ref(),
+                        context.quirks_mode
+                    ))
+                }
+                ref d => Cow::Borrowed(d)
+            };
+
+
             // When document colors are disabled, skip properties that are
             // marked as ignored in that mode, if they come from a UA or
```
Comment (upsuper) [inline components/style/properties/properties.mako.rs]: (Unrelated to this change, but I guess the "if" here should actually be "unless".)

File: components/style/properties/properties.mako.rs [hunk 3]
```diff
@@ -3336,10 +3314,42 @@ where
...
             // user style sheet.
-            if ignore_colors &&
+            if !device.use_document_colors() &&
```
Comment (upsuper) [inline components/style/properties/properties.mako.rs]: I think we should still use `ignore_colors` here, and I guess that we can move this check before variable resolution, although it may not be a big deal.

The check on `background-color` below should probably check `resolved_declaration` instead, but that can be in a separate issue. If that's the reason this if-block is placed after variable resolution... I'm probably okay with it.

File: components/style/properties/properties.mako.rs [hunk 4]
```diff
@@ -3336,10 +3314,42 @@ where
...
+                    Cow::Owned(unparsed.substitute_variables(
+                        id,
+                        context.builder.custom_properties.as_ref(),
+                        context.quirks_mode
+                    ))
+                }
+                ref d => Cow::Borrowed(d)
+            };
+
+
```
Comment (upsuper) [inline components/style/properties/properties.mako.rs]: Is this extra newline redundant?

**Review by upsuper — APPROVED** [review]
Generally looks good to me. r=me with the comments addressed.

File: components/style/properties/properties.mako.rs (hunk 1 above)
Comment (emilio) [reply components/style/properties/properties.mako.rs]: We only use it once now.

File: components/style/properties/properties.mako.rs (hunk 4 above)
Comment (emilio) [reply components/style/properties/properties.mako.rs]: Yes

File: components/style/properties/properties.mako.rs (hunk 1 above)
Comment (upsuper) [reply components/style/properties/properties.mako.rs]: Why? It seems to me we are still checking this for every declaration.

Comment (emilio) [reply components/style/properties/properties.mako.rs]: Oh, right... whoops. Will do that

File: components/style/properties/properties.mako.rs (hunk 2 above)
Comment (emilio) [reply components/style/properties/properties.mako.rs]: Yup, agreed, fixed.

File: components/style/properties/properties.mako.rs (hunk 3 above)
Comment (emilio) [reply components/style/properties/properties.mako.rs]: Yup, that's it... Nice catch, fixed :)
